scripts/scil_make_priors_functionnectome.py

```diff
@@ -0,0 +1,377 @@
+        vox_vol = np.zeros(templ_i.shape, dtype=np.float32)
+        for strmInd in vdict[vox]:
+            for v in stream_tupled[strmInd]:
+                vox_vol[v] = 1
+        save_tmp_map(vox_vol, out_Ftmp, templ_i.affine)
+
+# %%
+
+
+def main():
+    t0 = time.time()
+    t = t0
+    parser = _build_arg_parser()
+    args = parser.parse_args()
+
+    assert_inputs_exist(parser, [args.in_template_file], args.in_reg_dir)
+    if not os.path.isdir(args.in_dir_tractogram):
+        parser.error(f'{args.in_dir_tractogram} is not a directory.')
+    if not len(os.listdir(args.in_dir_tractogram)):
+        parser.error(f'{args.in_dir_tractogram} is empty.')
+    assert_output_dirs_exist_and_empty(parser, args, args.out_dir)
+
+    templ_i = nib.load(args.in_template_file)
+
+    if args.in_reg_dir is None:
+        templ_v = templ_i.get_fdata().astype(bool)
+        max_file_nb = templ_v.sum()
+    else:
+        list_reg_files = glob.glob(os.path.join(args.in_reg_dir, '*.nii*'))
+
+    if args.parallel_proc is not None:
+        nb_proc = int(args.parallel_proc)
+    else:
+        nb_proc = 1
+
+    trk_list = glob.glob(os.path.join(args.in_dir_tractogram, '*.trk'))
+    for n, trk_F in enumerate(trk_list):
+        print(os.path.basename(trk_F) + f' ({n+1}/{len(trk_list)})')
+        print('Loading the tractogram...')
+        trk_sft = load_tractogram(trk_F, templ_i, bbox_valid_check=False)
+        print('Filtering')
+        trk_sft = filter_streamlines_by_length(trk_sft, 25, 250)
+        trk_sft = hack_invalid_streamlines(trk_sft)
+        trk_sft.to_vox()
+        if args.in_reg_dir is None:
+            # For voxel-wise priors, "voxelizes" the tractograms
+            trk_sft = voxelize_tractogram(trk_sft)
+        tractomask = np.where(compute_tract_counts_map(trk_sft.streamlines, templ_i.shape),
+                              True, False)
+        empty_vol = templ_v > tractomask  # Voxels in templ but not in the tractogram
+        templ_v = templ_v*tractomask  # Removing voxels with no streamlines
+        if args.in_reg_dir is None:
+            voxel_list = np.argwhere(templ_v)
+            voxel_list = [tuple(ind) for ind in voxel_list]
+            stream_tupled = tupled_streamlines(trk_sft)
+            vox_dict = loop_on_strm(stream_tupled, voxel_list)
+            if nb_proc == 1:
+                print('Starting process')
+                strm_multi(vox_dict, stream_tupled, args.out_dir, templ_i)
+            else:  # Parallelize
+                vox_batchs = np.array_split(list(vox_dict), nb_proc)
+                vox_dict_batchs = [{tuple(v): vox_dict[tuple(v)] for v in vb} for vb in vox_batchs]
+                print('Starting parallel process')
+                with Pool(processes=nb_proc) as pool:
+                    poolCheck = pool.starmap_async(strm_multi, zip(vox_dict_batchs,
+                                                                   [stream_tupled]*nb_proc,
```

Comment:
Here instead of duplicate n times the data, prepare your batches before and give an array of batches with 1 batch per process. In the end you will just have 1 time the data in RAM instead of n times.

Reply:
normally, [stream_tupled]*nb_proc does not make copy of the obejct (each item of the list points to the same original object). And I can't split stream_tupled into batches, because each process may need any streamline in it. Or am I missing something?